src/test/groovy/unit/org/pih/warehouse/api/ProductApiTest.groovy

```diff
@@ -0,0 +1,178 @@
+package unit.org.pih.warehouse.api
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import groovy.transform.builder.Builder
+import io.restassured.RestAssured
+import io.restassured.builder.RequestSpecBuilder
+import io.restassured.http.ContentType
+import io.restassured.http.Cookie
+import io.restassured.path.json.JsonPath
+import io.restassured.specification.RequestSpecification
+import lombok.AllArgsConstructor
+import lombok.Data
+import lombok.NoArgsConstructor
+import org.grails.web.json.JSONObject
+import spock.lang.Shared
+import spock.lang.Specification
+import static io.restassured.RestAssured.*
+import io.restassured.response.Response
+import static org.hamcrest.MatcherAssert.assertThat
+import static org.hamcrest.Matchers.is
+import static org.hamcrest.Matchers.notNullValue
+import static org.hamcrest.Matchers.lessThanOrEqualTo
+
+class ProductApiTest extends Specification {
+
+    @Shared
+    String baseURI = System.getProperty("server.host")?:"http://localhost"
+
+    @Shared
+    String basePath = System.getProperty("server.path")?:"/openboxes"
+
+    @Shared
+    int port = System.getProperty("server.port")?Integer.valueOf(System.getProperty("server.port")):8080
+
+    @Shared
+    String username = System.getProperty("username")?:"admin"
+
+    @Shared
+    String password = System.getProperty("password")?:"password"
+
+    @Shared
+    String productId
+
+    @Shared
+    RequestSpecification requestSpecification
+
+    // Executed before the first spock test is executed
+    void setupSpec() {
+
+        RestAssured.basePath = basePath
+        RestAssured.baseURI = baseURI
+        RestAssured.port = port
+
+        JSONObject jsonObject = new JSONObject().
+                put("username", username).
+                put("password", password).
+                put("location", 1)
```

Comment:
Those dots shouldn't be at the beginning of the next lines?
```suggestion
        JSONObject jsonObject = new JSONObject()
                .put("username", username)
                .put("password", password)
                .put("location", 1)
```

Reply:
Hmm, I've never seen that syntax. Do you have a reference for that somewhere in Java / Groovy docs?

The rest assured docs provide a bunch of examples that use the syntax I used above.
https://github.com/rest-assured/rest-assured/wiki/Usage

And it looks like IntelliJ sometimes breaks when using the period at the beginning of the line (this is for a package name, but still makes me wary)
https://youtrack.jetbrains.com/issue/IDEA-242633/Groovy-code-formatting-incorrectly-wraps-qualified-new-expression-which-leads-to-compiler-error

Reply:
I've never seen that syntax 😅
I tried to search for breaking chain calls with the dot at the end, and I found a bunch of pages with this.
- https://unibeautify.com/docs/option-break-chained-methods
- https://github.com/prettier/prettier/issues/1565
- https://www.baeldung.com/java-config-spring-security
- I can see it also in the spring security official docs https://spring.io/guides/gs/securing-web/

Reply:
Ok, so there's some evidence for leading over trailing, so we should have a discussion about this. I guess I've mostly stayed to one line when chaining but I've only ever remembered seeing a trailing dot whenever I've seen chaining. I honestly don't recall ever seeing a leading dot. 

Interesting discussion in the Ruby https://github.com/standardrb/standard/issues/75 with mixed feelings.

Reply:
@kchelstowski @alannadolny Ok, after giving this a few days to marinate, I think I'm starting to warm up to the leading dot. I'm not 100% sold yet and I don't necessarily want to apply the line per method convention (like in the following example). But the leading dot does make it more readable.